kolibri/core/tasks/job.py

```diff
@@ -357,28 +359,17 @@ def enqueue_at(self, datetime, interval=0, repeat=0, args=(), kwargs={}):
 
     def _ready_job(self, *args, **kwargs):
         """
-        When self.validator is not None it runs the validator with
-        arguments passed to this method.
-
-        If validator raises an exception, we re-raise it otherwise we
-        return a job object.
+        Returns a job object with args and kwargs as its positional and keyword arguments.
         """
-        if self.validator is not None:
-            try:
-                validator_result = self.validator(*args, **kwargs)
-            except Exception as e:
-                raise e
-
-            kwargs["validator_result"] = validator_result
-
         job_obj = Job(
             self.func,
             *args,
             job_id=self.job_id,
             group=self.group,
             cancellable=self.cancellable,
             track_progress=self.track_progress,
+            extra_metadata=self.extra_metadata,
             **kwargs
         )
-
+        self.extra_metadata.clear()
```

Comment:
Not understanding the purpose of `self.extra_metadata` in this file - also, if multiple jobs are being created, there's a risk we are sharing the extra_metadata dict among them.

Given that this `self.extra_metadata` is always an empty object, it's not doing anything additional over the default (and safer) setting of it inside the Job constructor: https://github.com/learningequality/kolibri/blob/develop/kolibri/core/tasks/job.py#L189

Reply:
`self.extrametadata` plays a very important role there.

If a task function wants to set extra metadata, it can do that in its validator by doing `task_func.extra_metadata={"some": "thing"}`. This `extra_metadata` is binded to the registered job instance's `extra_metadata`, which gets passed to the job constructor via `_ready_job`. 

Every job that is registered will have its own `RegisteredJob.extra_metadata` so there's no chance of metadata being shared wrongly between multiple jobs.

Have a look at the demo `tasks.py` module to see it practically.

Reply:
I missed that in my initial review - but the way this is being used inside the validator is concerning.

In the last PR, there was some clean up to ensure that we were not using the registered task object as a transitory state object during task registration. This PR has reintroduced some of that concerning use again.

As a general rule, the validator functions should be pure functions without side effects - if we need to set the `extra_metadata` property - it seems that returning `extra_metadata` as a key on the dict that gets returned from the validator function is the appropriate way to do this, not by assigning it to the global registered task object.

Again, the issue here is if two tasks are being enqueued at the same time, with shared memory at play, then the use of the global object could lead either to the extra_metadata for one queuing being applied to another, or the extra metadata for one being cleared while another is being queued. It's a recipe for hard to narrow down bugs.

Reply:
The issue I can see is not with two different tasks because they'll have their own registered job instance so no issues with shared memory there. But yes if the same task is being enqueued by two different threads, then there can be race conditions with this approach. Is my thinking correct sir @rtibbles?

Reply:
Yes, it's the potential sharing of state between two tasks of the same type that I am concerned about.

Reply:
Will be fixed. :rocket: 

From now on, I'll keep this lesson in mind.